src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PETypeParameterSymbol.cs

```diff
@@ -235,6 +236,147 @@ private ImmutableArray<TypeWithAnnotations> GetDeclaredConstraintTypes()
+            }
+
+            bool? isNotNullableIfReferenceType = typeParameter.CalculateIsNotNullableIfReferenceType(inProgress, out isNonNullableValueType);
+
+            if (isNonNullableValueType)
+            {
+                Debug.Assert(isNotNullableIfReferenceType == true);
+                return true;
+            }
+
+            if (constraintType.NullableAnnotation.IsAnnotated() || isNotNullableIfReferenceType == false)
+            {
+                return false;
+            }
+            else if (constraintType.NullableAnnotation.IsOblivious() || isNotNullableIfReferenceType == null)
+            {
+                return null;
+            }
+
+            return true;
+        }
+
+        private bool? CalculateIsNotNullableIfReferenceType(ConsList<PETypeParameterSymbol> inProgress, out bool isNonNullableValueType)
+        {
+            if (inProgress.ContainsReference(this))
+            {
+                isNonNullableValueType = false;
+                return false;
+            }
+
+            if (this.HasValueTypeConstraint)
+            {
+                isNonNullableValueType = true;
+                return true;
+            }
+
+            bool? fromNonTypeConstraints = CalculateIsNotNullableFromNonTypeConstraints();
+
+            ImmutableArray<TypeWithAnnotations> constraintTypes = this.GetDeclaredConstraintTypes(inProgress);
+
+            if (constraintTypes.IsEmpty)
+            {
+                isNonNullableValueType = false;
+                return fromNonTypeConstraints;
+            }
+
+            bool? fromTypes = IsNotNullableIfReferenceTypeFromConstraintTypes(constraintTypes, inProgress, out isNonNullableValueType);
+
+            if (isNonNullableValueType)
+            {
+                Debug.Assert(fromTypes == true);
+                return true;
+            }
+
+            if (fromTypes == true || fromNonTypeConstraints == false)
+            {
+                return fromTypes;
+            }
+
+            Debug.Assert(fromNonTypeConstraints == null || fromNonTypeConstraints == true);
+            Debug.Assert(fromTypes != true);
+            return fromNonTypeConstraints;
+        }
+
+        private static bool? IsNotNullableIfReferenceTypeFromConstraintTypes(ImmutableArray<TypeWithAnnotations> constraintTypes, ConsList<PETypeParameterSymbol> inProgress, out bool isNonNullableValueType)
+        {
+            Debug.Assert(!constraintTypes.IsDefaultOrEmpty);
+
+            isNonNullableValueType = false;
+            bool? result = false;
+            foreach (TypeWithAnnotations constraintType in constraintTypes)
+            {
+                bool? fromType = IsNotNullableIfReferenceTypeFromConstraintType(constraintType, inProgress, out isNonNullableValueType);
+
+                if (isNonNullableValueType)
+                {
+                    Debug.Assert(fromType == true);
+                    return true;
+                }
+
+                if (fromType == true)
```

Comment:
>if (fromType == true) [](start = 16, length = 21)

nit: same comment here (having a method to pick strictest from two `bool?` or maybe dealing with `NullableAnnotation`)